Pluggable exchange

lzach opened the pull request:
Based on discussions with @BertR  and @perllaghu, I've created a pluggable version of the exchange, to allow custom exchanges for people who don't want to/can't use the default one. See #1130 for a bit more information.

The exchange file-tree has changed as follows:

<pre>
exchange/
├── abc/
│   ├── collect.py
│   ├── ...
│   └── submit.py
├── default/
│   ├── collect.py
│   ├── ...
│   └── submit.py
└── exchange_factory.py
</pre>

Here abc/ contains  abstract classes for each different exchange service (they aren't literally abstract, as the ABC meta-class conflicts with the LoggingConfigurable meta-class). The default/ folder contains the default implementations for each exchange service, which is just the old implementation.

The exchange_factory.py file contains the ExchangeFactory class, which is the interface for creating pluggable exchange services. It uses the Type traitlet for class variables for each exchange service, which is set to be a subclass of the corresponding class in abc/ and defaults to the corresponding class in  default/.

The ExchangeFactory class has instance methods that instantiate each of the exchange services. The new way of constructing the exchange classes would look something like:

<pre>
# lister = ExchangeLister(...) # old way
exchange = ExchangeFactory(config=config) # or parent=self, or similar
lister = exchange.List(...)
</pre>

For the apps, I've created an instance variable in the NbGrader app for the ExchangeFactory class, so that subclasses can use self.exchange.<exchange service> to create the services. The ExchangeFactory class gets initiated and configured in NbGrader.start, so before that method has been called, subclasses can't access the self.exchange variable. (Note, I removed line 39 in dbapp.py as it was creating a ExchangeList object and assigning it to an unused variable. Since this was done before the call to NbGrader.start it couldn't use the self.exchange, and as it was unused anyway, I figured I'd just delete the line.)

I also updated server_extensions/assignment_list/handlers.py to use the new pluggable exchange.

Finally, I should note that the method names in the ExchangeFactory class do not conform to PEP8. I did this to keep the names as close to the previous names as possible. If this is an issue, it should not be too much of a problem to change the names at this point.

Hope this all made sense.

jhamrick commented:
Thanks a lot, this is super exciting! Sorry I've been slow to get to reviews; I will try to find some time to dig through this in the next few days.

jhamrick commented on nbgrader/apps/api.py:
```diff
@@ -6,12 +6,12 @@
 from ..coursedir import CourseDirectory
 from ..converters import GenerateAssignment, Autograde, GenerateFeedback
-from ..exchange import ExchangeList, ExchangeReleaseAssignment, ExchangeReleaseFeedback, ExchangeFetchFeedback, ExchangeCollect, ExchangeError, ExchangeSubmit
+from ..exchange import ExchangeFactory, ExchangeError # ExchangeList, ExchangeReleaseAssignment, ExchangeReleaseFeedback, ExchangeFetchFeedback, ExchangeCollect, ExchangeError, ExchangeSubmit
```
Could you remove the commented code here?

jhamrick commented on nbgrader/exchange/abc/exchange.py:
```diff
@@ -0,0 +1,134 @@
+    pass
+
+
+class Exchange(LoggingConfigurable):
```
It seems like some of the traits and methods of this class make some strong assumptions about the files being located in a unix-like filesystem (e.g. the 'root') traitlet. Perhaps it would better to move these to the default exchange implementation rather than having them in the base class?

jhamrick replied:
Yes, that's a good point. An easy way to get around this would be to list the options for the default exchange classes in the `NbGrader` app. Specifically, in `apps/baseapp.py`, you can have:

```
        # include all the exchange actions
        for ex_name in exchange.default.__all__:
            ex = getattr(exchange.default, ex_name)
            if hasattr(ex, "class_traits") and ex.class_traits(config=True):
                classes.append(ex)
```

on line 153 (i.e. instead of importing the classes from `exchange` you import them from `exchange.default`).

This will of course though not help if you are using a different exchange than the default one. But in that case, perhaps we can assume that since it is a custom thing you would be familiar with the available options?

jhamrick commented on pytest.ini:
```diff
@@ -1,4 +1,4 @@
 [pytest]
-addopts = --strict-markers
+;addopts = --strict-markers
```
Looks like this unrelated change snuck in here?

jhamrick commented:
Oh actually I realized the tests do not explicitly create the exchange objects, so no work actually to do there!

lzach commented:
For some reason the travis build for nbextensions fails, even though it passes on my machine (both are using python 3.6). @BertR is checking whether it passes on his machine to determine whether it's an issue with my setup or something else.

lzach commented:
It appears that there's something different with my setup... trying to figure out what.

lzach commented:
Turns out that it was some changes that had been pushed to master between me pushing the original and pushing the requested changes that broke the tests. Fixed now.

lzach commented:
I've added in the code to pull the configuration from the default package (it also takes from the abc package, as there are some attributes that ought to be common to all implementations). I also added some documentation for how to write exchange plugins. It was originally written by @perllaghu, but I added in the new bits. Let me know if you want any changes to this.

The windows test fail for some reason and I can't check why as I don't have a windows machine.

moorepants commented:
Hi folks, a student group working with me has created ngshare:

https://github.com/lxylxy123456/ngshare

This allows nbgrader to be used on clusters and they've shown a very nice demo. But, it relies on this pull request. They have about 4 more weeks before school lets out and can spend some resources on getting this PR merged.

What can we do to help here?

I'm cc/ing the students: @Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998.

perllaghu commented:
> Hi folks, a student group working with me has created ngshare:
> 
> https://github.com/lxylxy123456/ngshare
> 
> This allows nbgrader to be used on clusters and they've shown a very nice demo. But, it relies on this pull request. They have about 4 more weeks before school lets out and can spend some resources on getting this PR merged.
> 
> What can we do to help here?
> 
> I'm cc/ing the students: @Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998.

It's failing on python 3.5 & 3.6 builds (possibly because we've used `f strings` a lot) - which then fails the azure.pipeline.

It would be interesting to see what else fails under 3.5 & 3.6

.... but, other than that - we're just waiting on a maintainer to do a final review & merge.

moorepants commented:
@jhamrick Are you the only maintainer that can approve merges? What can we do to get this one through the pipeline?

> It's failing on python 3.5 & 3.6 builds (possibly because we've used f strings a lot) - which then fails the azure.pipeline.

My guess, at this point in time, they will want to support Python 3.6+, so the fstrings will likely have to be removed.

moorepants commented:
@Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998 I think it would be worth it if you all reviewed this PR. Having more eyes on it could help @jhamrick feel comfortable to merge. Will you all review?

moorepants commented on nbgrader/apps/api.py:
```diff
@@ -48,7 +49,7 @@ def _log_level_changed(self, change):
         self.log.setLevel(new)
 
-    def __init__(self, coursedir=None, authenticator=None, **kwargs):
+    def __init__(self, coursedir=None, authenticator=None, exchange=None, **kwargs):
```
Needs an addition to the docstring explaining the new kwarg.

lzach replied:
Fixed this on my local branch.

moorepants commented on nbgrader/exchange/default/fetch_feedback.py:
```diff
@@ -2,11 +2,13 @@
-from .exchange import Exchange
-from ..utils import check_mode, notebook_hash, make_unique_key, get_username
+from nbgrader.exchange.abc import ExchangeFetchFeedback as ABCExchangeFetchFeedback
+from nbgrader.exchange.default import Exchange
```
Why have all the relative imports been changed? They should work as was, no?

lzach replied:
It was a long time since I wrote this, so can't quite remember, but I assume it was just a side effect of changing the structure of the package. It shouldn't matter either way.

Lawrence37 commented on nbgrader/exchange/abc/list.py:
```diff
@@ -0,0 +1,36 @@
+    cached = Bool(False, help="List assignments in submission cache.").tag(config=True)
+    remove = Bool(False, help="Remove, rather than list files.").tag(config=True)
+
+    def parse_assignment(self, assignment):
+        pass
+
+    def format_inbound_assignment(self, info):
+        pass
+
+    def format_outbound_assignment(self, info):
+        pass
+
+    def parse_assignments(self):
+        pass
```
I would remove these functions.

lzach replied:
Removed from my local branch.

Lawrence37 commented on nbgrader/exchange/abc/list.py:
```diff
@@ -0,0 +1,36 @@
+    def parse_assignments(self):
+        pass
+
+    def list_files(self):
+        pass
+
+    def remove_files(self):
+        pass
```
Perhaps implement the `start` function, which will ensure `inbound` and `cached` are not simultaneously set and call `list_files` or `remove_files` depending on the value of `remove`. Moving the `start` function from default.ExchangeList here would suffice.

I suggest documenting these two functions and making them raise a NotImplementedError, similar to `init_src` and `init_dest` in default.Exchange.

lzach replied:
Sounds reasonable and I've made this change in my local branch.

Lawrence37 commented on nbgrader/docs/source/exchange/exchange_api.rst:
```diff
@@ -0,0 +1,389 @@
+- nbgrader functionality requires a file called ``timestamp.txt`` to be in the submission, containing the timestamp of that submission. The creation of this file is the responsibility of this class.
+- Whilst nothing is done *as yet*, the default exchange checks the names of submitted notebooks, and logs differences.
+- Submissions need to record ``student_id``, as well as ``course_id`` & ``assignment_id``
+- The default exchange copies files to both an ``inbound`` and ``cache`` store. This may be significant considering ``ExchangeList``
```
Did the last sentence get cut off? If not, explain why it is significant.

lzach replied:
I have no idea what happened to that sentence. I've removed it for now as I'm not sure what is supposed to be significant here.

perllaghu replied:
The `cached` flag is used in base `nbgrader` to get the list of files in the submission cache (see https://github.com/jupyter/nbgrader/blob/dd0d66a41685c911971b32cbc3243659b2813a39/nbgrader/exchange/list.py#L20)

moorepants commented on nbgrader/exchange/abc/collect.py:
```diff
@@ -0,0 +1,26 @@
+    ).tag(config=True)
+
+
+    def init_src(self):
+        pass
+
+    def init_dest(self):
+        pass
+
+    def copy_files(self):
+        pass
```
Maybe it's designed as an abstract base class?

moorepants replied:
They could probably be removed though.

lzach replied:
I left some original structure in as I wanted feedback on what people thought would be useful to keep. These methods are definitely not needed though, and I've removed them in my local branch.

perllaghu replied:
If we follow the logic of the file-based exchange - all common methods are implemented in the base `Exchange` class, with `raise NotImplementedError`, and then defined in the actual implementation classes (with a `pass` if they do nothing.

This would be good login to follow if we want to make this code as usable as possible for others.

moorepants commented:
@Lawrence37 Thanks! The feedback looks good.

moorepants commented:
@lzach and @perllaghu Will you all have any cycles to update this PR? Or would you like us to open a new PR that builds on this? We could do either.

aalmanza1998 commented on nbgrader/apps/dbapp.py:
```diff
@@ -36,7 +36,6 @@ class DbBaseApp(NbGrader):
 
     def start(self):
         if sys.platform != 'win32':
-            lister = ExchangeList(coursedir=self.coursedir, parent=self)
```
Why was 'from ..exchange import ExchangeList' left in the imports if this line was removed?

lzach replied:
An oversight on my part, I've removed all unused imports I could find that relates to the exchange package now.

aalmanza1998 commented on nbgrader/docs/source/exchange/exchange_plugin.rst:
```diff
@@ -0,0 +1,77 @@
+    │   └── submit.py
+    └── exchange_factory.py
+
+The ``exchange.abc`` package contain all the Abstract Base Classes that custom exchange packages need to implement.
```
small typo, should say **contains** instead of **contain**

lzach replied:
Thanks, fixed.

aalmanza1998 commented on nbgrader/exchange/abc/release_feedback.py:
```diff
@@ -0,0 +1,5 @@
+
+
+class ExchangeReleaseFeedback(Exchange):
+    pass
```
missing newline at the end of file

lzach replied:
Fixed

aalmanza1998 commented on nbgrader/exchange/default/__init__.py:
```diff
@@ -0,0 +1,24 @@
+    "ExchangeReleaseAssignment",
+    "ExchangeReleaseFeedback",
+    "ExchangeSubmit"
+]
```
missing newline at the end of file

lzach replied:
fixed.

aalmanza1998 commented on nbgrader/server_extensions/assignment_list/handlers.py:
```diff
@@ -14,9 +14,9 @@
+from ...exchange import ExchangeFactory, ExchangeError
 from ...apps import NbGrader
 from ...coursedir import CourseDirectory
-from ...exchange import ExchangeList, ExchangeFetchAssignment, ExchangeFetchFeedback, ExchangeSubmit, ExchangeError
```
There is also an import for ExchangeList [here](https://github.com/lzach/nbgrader/blob/fe3ef73cdf22f79f8ca50ea4511366aca4e74ad3/nbgrader/server_extensions/formgrader/apihandlers.py#L8)

lzach replied:
Removed.

perllaghu commented on nbgrader/exchange/abc/submit.py:
```diff
@@ -0,0 +1,29 @@
+        )
+    ).tag(config=True)
+
+    add_random_string = Bool(
+        True,
+        help=dedent(
+            "Whether to add a random string on the end of the submission."
+        )
+    ).tag(config=True)
```
The "Everyone has read access" of standard nbgrader is how the instructor [just another user on the system] can read submissions copied into the directory by a student  [just another user on the system]

lzach replied:
You're probably right that this is implementation specific and I've moved this option to default.ExchangeSubmit in my own branch.

perllaghu replied:
If this is an nbgrader config option, we need to be sure the system doesn't break if a user has it defined in their config, and switch between the default file-based exchange and some external exchange.

lzach commented on nbgrader/apps/api.py:
```diff
@@ -92,7 +102,7 @@ def __init__(self, coursedir=None, authenticator=None, **kwargs):
         else:
             self.course_id = ''
-            self.exchange = ''
+            self.exchange_root = ''
```
Well spotted, I've fixed this on my local branch.

lzach commented on nbgrader/exchange/abc/fetch_assignment.py:
```diff
@@ -0,0 +1,16 @@
+
+    replace_missing_files = Bool(False, help="Whether to replace missing files on fetch").tag(config=True)
+
+    def copy_if_missing(self, src, dest, ignore=None):
+        pass
```
You're probably right, I've removed it from my local branch as it doesn't do anything anyway right now.

lzach commented on nbgrader/exchange/abc/list.py:
```diff
@@ -0,0 +1,36 @@
+from .exchange import Exchange
+
+
+def _checksum(path):
+    m = hashlib.md5()
+    m.update(open(path, 'rb').read())
+    return m.hexdigest()
```
Removed from local branch.

lzach commented on nbgrader/exchange/default/fetch.py:
```diff
@@ -0,0 +1,16 @@
+
+    def __init__(self, *args, **kwargs):
+        super(ExchangeFetch, self).__init__(*args, **kwargs)
+        msg = (
+            "The ExchangeFetch class is now deprecated, please use "
+            "ExchangeFetchAssignment instead. This class will be removed in a "
+            "future version of nbgrader.")
+        warnings.warn(msg, DeprecationWarning)
+        self.log.warning(msg)
```
I've completely removed the default/fetch.py and default/release.py files, as they can't be instantiated using the factory anyway.

lzach commented on nbgrader/exchange/abc/submit.py:
```diff
@@ -0,0 +1,29 @@
+        )
+    ).tag(config=True)
+
+    def init_release(self):
+        pass
+
+    def check_filename_diff(self):
+        pass
```
You're probably right, I've removed them from my branch.

moorepants commented:
Thanks for the updates @lzach!

Does anyone know if @jhamrick is the only person with merge rights on this repo?

lzach commented:
@moorepants If it's alright with you, I'll just push the fixes to this PR?

As far as I know, @jhamrick  is indeed the only one, though I'm not 100% sure.

moorepants commented:
> If it's alright with you, I'll just push the fixes to this PR?

Definitely!

moorepants commented:
We need to figure out the CI build failures too.

moorepants commented:
Also, isn't this set of changes going to need unit tests? I didn't see any files labeled as tests on first glance.

perllaghu commented:
Switch away from fstrings too

On Tue, 5 May 2020, 10:12 pm Jason K. Moore, <notifications@github.com>
wrote:

> We need to figure out the CI build failures too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/nbgrader/pull/1238#issuecomment-624308875>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALJCDBXHMMPL67BJR4VZUDRQB6LTANCNFSM4I6DKCCA>
> .
>

lzach commented:
I'm currently writing tests for our version of the exchange, and will add in tests here as well. Almost all the code is covered by previous tests, as this mostly only changes the structure of the exchange package, whereas most of the logic remains exactly the same.

@perllaghu I can't find any f-strings in my nbgrader branch.  I'll run tests on all supported versions of python before I push, so hopefully I'll catch whatever is causing the checks to fail.

perllaghu commented:
Ok.... Something else was making it fail the 3.5 & 3.6 environments
then....


But looking good!

On Tue, 5 May 2020, 10:22 pm Ludvig Kihlman, <notifications@github.com>
wrote:

> @perllaghu <https://github.com/perllaghu> I can't find any f-strings in
> my nbgrader branch. I'll run tests on all supported versions of python
> before I push, so hopefully I'll catch whatever is causing the checks to
> fail.
>

moorepants commented:
@Lawrence37, @lxylxy123456, @rkevin-arch, @aalmanza1998 You all can also help by looking into the CI errors.

lzach commented:
I've pushed the latest changes. All tests passes except for the mypy ones. This seems to a problem on all pull requests though, so not sure what can be done about it. Mypy doesn't seem to like the demos folder, as far as I can tell.

lzach commented:
Pushed a fix for the failing mypy tests as described by @BertR in #1328

willingc commented:
@jhamrick I would recommend dropping Python 3.5 now since it goes EOL in a few months and has been in security fix mode since last year. Dropping it would allow f-strings and better asyncio support.
